articles/zero_copy.md

```diff
@@ -0,0 +1,249 @@
+One additional publishing use case is allowing the user to loan the message to the middleware during asynchronous publishing.
+This use case comes up when all or part of the data being published is located in memory that the middleware cannot allocate from, e.g. a hardware buffer via memory mapped I/O or something similar, and the user wants to still have zero copy and asynchronous publishing.
+In this case, the user needs to call publish, then keep the message instance immutable until the middleware lets the user know that it is done with the loaned data.
+
+This is a narrow use case and will require additional interfaces to support, therefore it will be out of scope for this document, but it is mentioned for completion.
+
+### Subscription Use Cases
+
+There are two ways the users may take message instances from a subscription when data is available:
+
+1. Taking direction from the `Subscription` after polling it for data availability or waiting via a wait set.
+2. Using an `Executor`, which takes the data from the user and delivers it via a user-defined callback.
+
+Note: It is assumed that the user will be abe to take multiple messages at a time if they are available.
+
+In the first case, the user could choose to either:
+
+* Manage the memory for the message instance themselves, providing a reference to it, into which the middleware should fill the data.
+* Take one or more loaned messages from the middleware and return the loans later.
+
+In the second case, the user is delegating the memory management ot he client library via the `Executor`.The `Executor` may or may not borrow data from the middleweare, but the user callback does not care, so this can be considered an implementation detail.
+The user should be able to influence what the `Executor` does, and in the case that memory needs to be allocated, the user should be able to provide an allocator or memory management strategy which would influence the `Executor`'s behavior.
+
+## Requirements
+
+Based on the the use cases above, the general requirements are as follows:
+
+* Users must be able to avoid all memory operations and copies in at least one configuration.
+
+
+### Publisher Requirements
+
+* The user must be able to publish from messages allocated in their stack or heap.
+* The user must be able to get a loaned message, use it, and return it during publication.
+* When taking a loan, the middleware should not do anything "special", that is that the user must be able to influence the allocation.
+
+### Subscription Requirements
+
+The following requirements should hold whether the user is polling or using a wait set.
+
+* The user must be able to have the middleware fill data into messages allocated in the user's stack or heap.
+* The user must be able to get a loaned message from the middleware when calling take.
+* The user must be able to get a sequence of loaned messages from the middleware when calling take.
+* The loaned message or sequence must be returned by the user.
```

Comment:
Note from @Karsten1987:

> We should specify how we can make sure that the user doesn't hold too many of these loaned messages when not taking them via the executor/callback but rather via polling the waitset.

> AUTOSAR specifies this as not possible to gain access to the sequence without specifying an according callback where the sequence eventually gets passed in via std::unique_ptr.

Reply:
@mjcarroll: How does the AUTOSAR approach help? I still can just keep the `unique_ptr` in my container forever.

Reply:
I don't think that AUTOSAR helps, this was from our initial brainstorming notes.  The important point was that if users hold on to loaned messages, at some point the underlying mechanism is going to stop working.